src/Allowlist.sol

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable/access/Ownable2StepUpgradeable.sol";
+
+/// @notice Minimal external interface for ZKMe-like approval checker.
+interface IZkMe {
+    function hasApproved(
+        address cooperator,
+        address user
+    ) external view returns (bool);
+}
+
+/// @notice Simple allowlist contract with optional ZKMe fallback, adapted from the Rust implementation.
+contract Allowlist is Initializable, UUPSUpgradeable, Ownable2StepUpgradeable {
+    struct ZkMeSettings {
+        address contractAddr;
+        address cooperator;
+        bool enabled;
+    }
+
+    struct AllowlistStorage {
+        mapping(address => bool) allowed;
+        ZkMeSettings zkMe;
+    }
+
+    // keccak256(abi.encode(uint256(keccak256("maxbtc.allowlist.storage")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant ALLOWLIST_STORAGE_SLOT =
+        0xa72203c0a1e9d976a073cbb6bf59891bede3e473d11183dd860e5c4524e5e600;
+
+    event AddressAllowed(address indexed account);
+    event AddressDenied(address indexed account);
+    event ZkMeSettingsUpdated(
+        address contractAddr,
+        address cooperator,
+        bool enabled
+    );
+    event ZkMeSettingsReset();
+
+    error InvalidAddress();
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address owner_) public initializer {
+        if (owner_ == address(0)) {
+            revert InvalidAddress();
+        }
```

Comment:
This check is already performed by `Ownable`, there is no need to reimplement it here.